### ISSUE TITLE
Shadow LOD displaying as lowest LOD when Model Detail is set to Medium or Low

### ISSUE THREAD

**Issue by GrantCL** [Facepunch/garrysmod-issues#2953]
Hello, this is the first time I've reported an issue to the GMod bug tracker. Forgive me if I break protocol.

I have created several model addons for GMod, mostly ragdolls and props. Recently, it was brought to my attention that one of my models was displaying incorrectly ingame. Instead of the ragdoll model, it would render a flat circle instead.

### Details
I am very adamant about providing LOD optimizations for my addons, and it appears that my subscribers are now experiencing an odd issue with models that use the Shadow LOD command, and the Model Detail setting.

To illustrate the issue:
http://images.akamai.steamusercontent.com/ugc/102851689556595393/8B4830F17EAEA919106ACE7810E9267CA3B8F95E/

The model shown uses a special Shadow LOD which overrides its shadow, forcing the shadow to render as a flat circle instead of the original model. I did this in homage to the old N64 style of rendering shadows. Whenever the Model Detail setting is set to Medium or Low, GMod falls back to the Shadow LOD rather than the lowest LOD for this model.

Strangely, this issue was undetectable for me up until now, as I had never defined a Shadow LOD that was not explicitly the original model. This issue seemed like unintended behavior(why would you ever want to display the Shadow LOD to the user?), which is why I'm reporting it.

To help troubleshoot the issue, I made several different versions of the same model with a differing number of LODs. I used Alyx and other HL2/TF2 ragdolls/props as a control, and they seem to function correctly when throttling the Model Detail setting. They do not use their Shadow LOD, but rather their 3rd/4th LOD(of 6 LODs) as their lowest when viewing them at point blank. Addon-related models and ragdolls fall back to the Shadow LOD whenever it is available(explicitly defined in the QC), regardless of how many regular LODs they possess.

Models that do not have a Shadow LOD appear to function normally. Models that utilize LODs, but have no Shadow LOD, appear to function normally as well. I double-checked several HL2 and TF2 models to make sure they have Shadow LODs defined, and they do. I also tested this issue with wireframe enabled, so I could detect when the LOD switch occured.

I'm afraid I was only able to detect this issue while testing amongst my own addons. I wouldn't know where to begin tracking down other GMod addons outside my own that utilize a Shadow LOD.

Here is a link to the addon that illustrates the issue: http://steamcommunity.com/sharedfiles/filedetails/?id=808420593

GMod Version: 2016.10.29

### Steps to reproduce
-Spawn an addon-based ragdoll, prop, vehicle, or model that utilizes Shadow LODs explicitly defined in the QC. This can be from a spawnlist, or from a tab(NPC, Vehicle, etc).
-Change your Model Detail from High to Medium or Low.
-The Shadow LOD should display instead of the appropriate LOD for the model.
-Switching back to High Model Detail will restore the model to normal.

**Comment by willox** [comment]
Is this what the shadow LOD should look like?

![](http://files.facepunch.com/willox/2016-12-07/6ouyOwPT.png)

**Comment by GrantCL** [comment]
I was going to post the QC straight, but it turns out its a bit more massive than I remembered. Here is a zipped up version of the QC.

[super_scout.zip](https://github.com/Facepunch/garrysmod-issues/files/637406/super_scout.zip)

The Shadow LOD portion of this model is super simple.

-"lod_shadow.smd" is the circular disc that represents the shadow(I might change this to a sphere later for a better effect).

-"null" is just a tiny box at the model's origin, which I use to remove the hat from the shadow(removemesh crashes studiomdl, so I replaced the hat with a dummy model instead).

-I use a bonetreecollapse to collapse all bones to the root bone, found at the origin of the model. This ensures the disc isn't affected by any of the other bones.

-I suspected that "$mostlyopaque" might be mucking with the LODs, but Alyx uses the same command, and she works fine. Same goes for the some of the TF2 characters. 

Hope that helps, Thank you!

**Comment by willox** [comment]
I think the only way this could be properly fixed is in the model compiler itself. If you give me enough files to compile one of the models, I'll do that.

You could also fix it by having > 3 LODs or using `$allowrootlods 1` (where 1 is the amount of LODs you want to actually be visible)

**Comment by GrantCL** [comment]
After reading your suggestion, I gave it a few hours of experimentation to see if I could track down the exact point where it breaks. Your call on $allowrootlods was spot on, it fixes the model in every scenario. Even when there are no LODs other than the ShadowLOD. Crazy really, I had never heard of that command until now, and I would not have even fathomed it existed.

I also tested against differing numbers of LODs with and without the command. Testing between 1-6 LODs without the command, I found the ShadowLOD issue only presents itself when there is 1 LOD besides the ShadowLOD. It works at 2 or more LODs.

If this command is the correct way to handle this ShadowLOD error, then the fix should be on my end. Still, It seems strange that the ShadowLOD would ever be allowed to render directly to the user. I generally thought that your ShadowLOD could be as messy as you wanted it to be, since it's all about speeding up shadow processing. Knowing that my ShadowLODs could be exposed to the user, I'll have to be a bit more careful when testing my addons.

I suppose the scenario, "My shadow is a completely different model from the main model", doesn't come up often enough to warrant patching the entire system just for an edge case like this.

Thanks for the help! I'll leave it up to you to decide whether or not this is a fix worthy issue. In the meantime, I'll go through all my addons and patch them with $allowrootlods.


If you'd still like to mess with it, here is the source material. It includes all of my test LODs.
[superscout_src.zip](https://github.com/Facepunch/garrysmod-issues/files/640347/superscout_src.zip)
